grammar/item.lyg

```diff
@@ -84,20 +85,23 @@ ImplItemKind =
   | MacroCall:ItemMacroCall
   ;
 
-FnHeader = constness:"const"? unsafety:"unsafe"? asyncness:"async"? { "extern" abi:Abi }?;
-FnDecl = name:IDENT generics:Generics? "(" args:FnArgs? ")" { "->" ret_ty:Type }? where_clause:WhereClause?;
+FnHeader = constness:"const"? asyncness:"async"? unsafety:"unsafe"? { "extern" abi:Abi }?;
+FnDecl = name:IDENT generics:Generics? "(" params:FnParams? ")" { "->" ret_ty:Type }? where_clause:WhereClause?;
 
 // FIXME(eddyb) find a way to express this `A* B?` pattern better
-FnArgs =
-  | Regular:FnArg+ %% ","
-  | Variadic:"..."
-  | RegularAndVariadic:{ args:FnArg+ % "," "," "..." }
+FnParams =
+  | Regular:FnParam+ %% ","
+  | CVariadic:FnCVariadicParam
+  | RegularAndCVariadic:{ args:FnParam+ % "," "," c_variadic:FnCVariadicParam }
   ;
-FnArg =
+FnParam = attrs:OuterAttr* kind:FnParamKind;
+FnParamKind =
   | SelfValue:{ mutable:"mut"? "self" }
   | SelfRef:{ "&" lt:LIFETIME? mutable:"mut"? "self" }
```

Comment:
The `self` param is only allowed in the first parameter syntactically (Also, consider lifting this out to its own rule.)